server/src/main/java/org/apache/iotdb/db/mpp/memory/MemoryPool.java

```diff
@@ -79,11 +141,35 @@ public synchronized void free(String queryId, long bytes) {
     } else {
       queryMemoryReservations.put(queryId, queryReservedBytes);
     }
-
     reservedBytes -= bytes;
+
+    if (memoryReservationFutures.isEmpty()) {
+      return;
+    }
+    Iterator<MemoryReservationFuture<Void>> iterator = memoryReservationFutures.iterator();
+    while (iterator.hasNext()) {
+      MemoryReservationFuture<Void> future = iterator.next();
+
+      if (future.isCancelled()) {
+        iterator.remove();
+        continue;
+      }
+
+      long bytesToReserve = future.getBytes();
+      if (maxBytes - reservedBytes < bytesToReserve
+          || maxBytesPerQuery - queryMemoryReservations.getOrDefault(queryId, 0L)
```

Comment:
It seems that `MemoryReservationFuture` doesn't know which query it belongs to. In other words, the `future` and the `queryId` may represent different query. So we cannot ensure the queryMemoryReservations of the `future` does not exceed the `maxBytesPerQuery `

Reply:
Make sense. Will fix it.

Reply:
Fixed.